---
title: "RNA-seq analysis in R"
subtitle: "QC of raw reads with FastQC - Solutions"
---

### Exercise 1

> 1. 
>       a) Check the location of the current directory using the command `pwd`  
>       b) If the current directory is not
>       `/home/participant/Course_Materials`, then navigate to the
>       **Course_Materials** directory using the `cd` (**c**hange
>       **d**irectory) command:  
> `cd ~/Course_Materials`  
>
> 2.   
>       a) Use `ls` to list the contents of the directory. There should be directory
>    called **fastq**    
>       b) Use `ls` to list the contents of the **fastq** directory:  
>    `ls fastq`
>
> 3. Create a new directory for the QC results called **QC** using the `mkdir`
>    command:  
> `mkdir QC`
>
> 4. Run fastqc on one of our samples:  
> `fastqc fastq/SRR7657883.sra_1.fastq.gz`
>
> 5. The previous command has written the report to the **fastq** directory -
>    the default behaviour for fastqc. We want it in the **QC** directory. 
>       a) Use the `rm` (remove) command to delete the report:   
>          `rm fastq/SRR7657883.sra_1_fastqc.html`  
>       b) Also delete the associated zip file (this contains all the figures
>       and data tables for the report)

```
rm -f fastq/SRR7657883.sra_1_fastqc.zip
```

>
> 6. Run the FastQC again, but this time try to use the `-o` option to have
>    the report written to the **QC** directory.

```
fastqc -o QC fastq/SRR7657883.sra_1.fastq.gz
```

>  
> 7. Open the html report in a browser and see if you can answer these questions:  
>           A) What is the read length? **100**  
>           B) Does the quality score vary through the read length?  
>               Yes, the first few bases and the last few bases are typically
>               of lower quality.  
>           C) How is the data's quality?  
>               Overall, pretty good.
