---
title: "RNA-seq analysis in R"
subtitle: "QC of Aligned Reads - exercise solutions"
output: html_document
---

## 1. A quick look at the alignment metrics with `samtools`

### Exercise 1

> 1. Check your current working directory and if necessary navigate to 
>    the `Course_Materials/` directory using the command `cd` (change
>    directory).
>
> `pwd` - to check **p**resent **w**working **d**irectory
> 
> `cd ~/Course_Materials`  - if necessary
>
> 2. Use the `samtools flagstat` command to generate alignment metrics for
>    the sorted bam file you created in the previous practical.  
> ` samtools flagstat bam/SRR7657883.sorted.bam`
> ```
> 64434897 + 0 in total (QC-passed reads + QC-failed reads)
> 4193507 + 0 secondary
> 0 + 0 supplementary
> 0 + 0 duplicates
> 61244035 + 0 mapped (95.05% : N/A)
> 60241390 + 0 paired in sequencing
> 30120695 + 0 read1
> 30120695 + 0 read2
> 54707584 + 0 properly paired (90.81% : N/A)
> 55067498 + 0 with itself and mate mapped
> 1983030 + 0 singletons (3.29% : N/A)
> 161370 + 0 with mate mapped to a different chr
> 139571 + 0 with mate mapped to a different chr (mapQ>=5)
> ```
> Q) What percentage of the reads have aligned to the genome? - 95.05% of reads
> have been mapped to the reference genome.


## 2. More detailed metrics with Picard Tools

### 2.1 Duplication metrics

### Exercise 2.1

> 1. Run Picard's MarkDuplicates tool on the sorted bam file using the
>    following command:
```
java -jar picard/picard.jar MarkDuplicates \
     INPUT=bam/SRR7657883.sorted.chr1.bam \
     OUTPUT=bam/SRR7657883.mkdup.chr1.bam \
     METRICS_FILE=bam/SRR7657883.mkdup_metrics.chr1.txt \
     CREATE_INDEX=true
```
> **Note**: The `\` at the end of each line tells the terminal that when you press 
> `Enter`, you have not yet finished typing the command. You can if you wish, type
> the whole command on a single line, omitting the `\` - The command is written across
> multiple lines here just to make it easier to read.
>
> Q. What is the duplication rate for this bam file? 
> ~55%. 
> Note that although the column headers for Picard say "PERCENT" or "PCT" the 
> number is in fact the decimal fraction and need to be multiplied by 100 for 
> percent. Just an odd quirk of Picard.

### 2.2 Alignment metrics

### Exercise 2.2

> 1. Run Picard's `CollectAlignmentSummaryMetrics` tool on the chr1 sorted bam  
>    providing the following options.
>     * INPUT - The sorted bam file
>     * OUTPUT - "bam/SRR7657883.chr1.alignment_metrics.txt"
>     * REFERENCE_SEQUENCE - "references/Mus_musculus.GRCm38.dna_sm.primary_assembly.fa"
```
java -jar picard/picard.jar CollectAlignmentSummaryMetrics \
     INPUT=bam/SRR7657883.chr1.sorted.bam \
     OUTPUT=bam/SRR7657883.chr1.alignment_metrics.txt \
     REFERENCE_SEQUENCE=references/Mus_musculus.GRCm38.dna_sm.primary_assembly.fa
```


### 2.2 Insert Size metrics

### Exercise 2.2

> 1. Run Picard's `CollectInsertSizeMetrics` tool on the chr1 sorted bam 
>    providing the following options.
>     * INPUT - The sorted chr 1 only bam file
>     * OUTPUT - bam/SRR7657883.chr1.insert_size.txt
>     * HISTOGRAM_FILE - bam/SRR7657883.chr1.insert_size.pdf
>
```
java -jar picard/picard.jar CollectInsertSizeMetrics \
     INPUT=bam/SRR7657883.chr1.sorted.bam \
     OUTPUT=bam/SRR7657883.chr1.insert_size.txt \
     HISTOGRAM_FILE=bam/SRR7657883.chr1.insert_size.pdf
```

### 2.3 RNA alignment metrics

### Exercise 2.3

> 1. Run Picard's `CollectRnaSeqMetrics` tool on the sorted bam file providing
>    the following options:
>       * INPUT - The sorted bam file
>       * OUTPUT - "bam/SRR7657883.chr1.RNA_metrics.txt"
>       * REF_FLAT - the RefFlat reference file
>       * STRAND - "NONE"
```
java -jar picard/picard.jar CollectRnaSeqMetrics \
     INPUT=bam/SRR7657883.chr1.sorted.bam \
     REF_FLAT=references/Mus_musculus.GRCm38.102.txt \
     OUTPUT=bam/SRR7657883.chr1.RNA_metrics.txt \
     STRAND=NONE
```
>
> The results of this analysis are best viewed graphically, we will do this in
> the next exercise.

## 3. Visualising QC results with MultiQC

### Exercise 3.1

> 1. Run multiqc on the bam directory: 
```
multiqc -n Alignment_QC_Report.html -o bam bam`
```
>     * `-n` - a name for the report
>     * `-o` - the directory in which to place the report
> 2. Open the html report that was generated by multiqc and inspect the QC plots
```
xdg-open mutliqc_report.html
```


### Exercise 3.2

> In the `metrics` directory you should find Picard metrics for all of the 
> samples.
> 1. Run multiqc on the contents of the metrics directory.  
```
multiqc -z -n Alignment_QC_Report.html -o metrics metrics
```
> --> [Alignment_QC_Report.html](../additional_scripts/Alignment_QC_Report.html)  
> 2. Open the html report that was generated by multiqc and inspect the QC plots   
>
> Q. Are there any bam files that look problematic?
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX



