---
title: "Run through full analysis"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: 
    html_notebook:
         code_folding: hide
         toc: yes
bibliography: ref.bib
---


```{r setup, message = FALSE}
library(tximport)
library(DESeq2)
library(ggfortify)
library(tidyverse)
library(patchwork)
library(ggdendro)
```

## Details

Data is from:
Hu et al. (2020) Transcriptomic Profiling of Mouse Brain During Acute and 
Chronic Infections by Toxoplasma gondii Oocysts. Frontiers in Microbiology 11:2529   
		
https://www.frontiersin.org/article/10.3389/fmicb.2020.570903     

### ABSTRACT
Infection by the protozoan Toxoplasma gondii can have a devastating impact on the structure and function of the brain of the infected individuals, particularly immunocompromised patients. A systems biology view of the brain transcriptome can identify key molecular targets and pathways that mediate the neuropathogenesis of cerebral toxoplasmosis. Here, we performed transcriptomic analysis of the brain of mice infected by T. gondii Pru strain oocysts at 11 and 33 days post-infection (dpi) compared to uninfected (control) mice using RNA sequencing (RNA-seq). T. gondii altered the expression of 936 and 2,081 transcripts at 11 and 33 dpi, respectively, and most of these were upregulated in the infected brains. Gene Ontology (GO) enrichment and pathway analysis showed that immune response, such as interferon-gamma (IFN-Î³) responsive genes were strongly affected at 11dpi. Likewise, differentially expressed transcripts (DETs) related to T cell activation, cytokine production and immune cell proliferation were significantly altered at 33 dpi. Host-parasite interactome analysis showed that some DETs were involved in immune signaling, metabolism, biosynthesis-related processes and interspecies interaction. These findings should increase knowledge of the mouse brain transcriptome and the changes in transcriptional regulation and downstream signaling pathways during acute and chronic T. gondii infections.

SRA Accession: PRJNA483261

Data has been downloaded and quantified using Salmon.

```{r loadsamplesheet, message = FALSE}
# Read the sample information into a data frame
samplesheet <- read_tsv("data/samplesheet_corrected.tsv", col_types = c("ccff")) %>% 
  arrange(Group, TimePoint, Replicate)
samplesheet
```

## Reading in the count data

Count data is read from Salmon outputs and summarised to gene.

```{r readSalmon, message=FALSE}
tx2gene <- read_tsv("references/tx2gene.tsv", col_types = "cc")

# create the "files" object
files <- str_c("salmon/", samplesheet$SampleName, "/quant.sf")
files <- set_names(files, samplesheet$SampleName)

txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
```

```{r rawCounts}
rawCounts <- round(txi$counts, 0) 
keep <- rowSums(rawCounts) > 5
filtCounts <- rawCounts[keep,]
rlogcounts <- rlog(filtCounts)
```

## Correlation Matrix

Top 500 genes by variance only

```{r, fig.wdith=8, fig.height=8}
samsht <- samplesheet%>%   
  mutate(SampleGroup = str_c(Group, ".", TimePoint)) %>% 
  mutate(SampleID = str_c(SampleGroup, ".", Replicate)) %>% 
  select(SampleName, SampleGroup, SampleID)

topVar <- order(rowVars(rlogcounts), decreasing = TRUE)[1:500]
plotDat <- cor(rlogcounts[topVar,]) %>%
  as.data.frame() %>%
  rownames_to_column("Xs") %>%
  pivot_longer(names_to = "Ys", values_to = "Cor", -Xs) %>% 
  mutate(CorTxt =round(Cor, 3)) %>%
  left_join(samsht, by = c(Xs="SampleName"))  %>% 
  mutate(X = factor(SampleID)) %>% 
  select(-SampleGroup, -SampleID) %>%
  left_join(samsht, by = c(Ys="SampleName"))  %>% 
  mutate(Y = factor(SampleID))
    
ggplot(plotDat, aes(x = X, y = Y, fill = Cor)) +
    geom_tile(col = "grey") +
    geom_text(aes(label = CorTxt), size=4) +
    scale_fill_gradientn(colors = c("#FFFFFF", "#B90505"), 
                         breaks = seq(0, 1, 0.2),
                         limits = c(0.5,1)) +
    guides(fill = guide_colorbar(barheight = 10)) +
    theme(
        aspect.ratio = 1,
        axis.text.x = element_text(angle = 90, 
                                   hjust = 1, 
                                   vjust = 0.5, 
                                   size = 13),
        axis.text.y = element_text(size = 13),
        plot.title = element_text(hjust = 0.5),
        panel.background = element_blank())

```

## PCA

```{r pca, fig.width = 12, fig.height = 7}
pcDat <- t(rlogcounts) %>% prcomp()

p1 <- autoplot(pcDat,
               data = samplesheet, 
               colour = "Group", 
               shape = "TimePoint",
               size=5) +
        theme(legend.position="bottom")

p2 <- autoplot(pcDat,
               x = 2,
               y = 3,
               data = samplesheet,
               colour = "Group", 
               shape = "TimePoint",
               size=5) +
        theme(legend.position="bottom")

p1 + p2 +  plot_layout(guides = "collect") & theme(legend.position = 'bottom')

```


## Hierachical clustering

```{r, fig.width  =  12, fig.height = 7}
dendro.dat <- t(rlogcounts) %>%
   dist(method = "euclidean") %>%
   hclust() %>%
   as.dendrogram() %>%
   dendro_data()

labelDat <- dendro.dat$labels %>%
    mutate(SampleName = as.character(label)) %>%
    left_join(samplesheet, by = "SampleName") %>% 
    mutate(SampeID = str_c(Group, ".", TimePoint, ".", Replicate))

axisBreaks <- pretty(dendro.dat$segments$yend)[-1] %>% head(-1)

ggplot(dendro.dat$segment) +
    geom_segment(aes(x = x, y = y, xend = xend, yend = yend)) +
    geom_label(data = labelDat,
              aes(x = x,
                  y = y,
                  label = SampeID,
                  fill = Group),
                  hjust = 0,
                  nudge_y = 1) +
    guides(fill = FALSE) +
    labs(x = NULL, y = "Distance", title = NULL) +
    scale_y_reverse(expand = c(0.2, 0), breaks = axisBreaks) +
    coord_flip() +
    theme(axis.title.y = element_blank(),
          axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          plot.title = element_text(hjust = 0.5),
          panel.background = element_blank())
```

## Differential gene expression analysis - Simple Model (~Group)

Differential expression analysis is carried out using `DESeq2`.

Number of genes with padj < 0.05 for Test v Control:

```{r DGEsimple, message=FALSE}
simple.model <- as.formula(~ Group)

ddsObj.raw <- DESeqDataSetFromTximport(txi = txi,
                                      colData = samplesheet,
                                      design = simple.model)
keep <- rowSums(counts(ddsObj.raw)) > 5
ddsObj.filt <- ddsObj.raw[keep,]
ddsObj.simple <- DESeq(ddsObj.filt)
results.simple <- results(ddsObj.simple, alpha=0.05)
sum(results.simple$padj < 0.05, na.rm = TRUE)
```

## Differential gene expression analysis - Additive Model (~TimePoint + Group)

Number of genes with padj < 0.05 for Test v Control:

```{r, message=FALSE}
additive.model <- as.formula(~ TimePoint + Group)

ddsObj.raw <- DESeqDataSetFromTximport(txi = txi,
                                      colData = samplesheet,
                                      design = additive.model)
keep <- rowSums(counts(ddsObj.raw)) > 5
ddsObj.filt <- ddsObj.raw[keep,]
ddsObj.additive <- DESeq(ddsObj.filt)
results.additive <- results(ddsObj.additive, alpha=0.05)
sum(results.additive$padj < 0.05, na.rm = TRUE)
```

### Compare additive v simple using LRT

Number of genes with padj < 0.05 for LRT of '~ TimePoint + Group' vs '~ Group': 

```{r compareModels, message=FALSE}
# Compare the two designs
ddsObjC <- DESeq(ddsObj.additive, test="LRT", reduced=simple.model)
resCvCS <- results(ddsObjC)
sum(resCvCS$padj < 0.05, na.rm=TRUE)
```

This is not many genes for which the LRT suggests the additive model differs from
the simpel model.

## Interaction model

```{r, message=FALSE}
interaction.model <- as.formula(~ TimePoint * Group)

ddsObj.raw <- DESeqDataSetFromTximport(txi = txi,
                                      colData = samplesheet,
                                      design = interaction.model)
keep <- rowSums(counts(ddsObj.raw)) > 5
ddsObj.filt <- ddsObj.raw[keep,]
ddsObj.interaction <- DESeq(ddsObj.filt)
results.interaction.11 <- results(ddsObj.interaction, 
                                  name="Group_Test_vs_Control",
                                  alpha=0.05)
results.interaction.33 <- results(ddsObj.interaction, 
          contrast = list(c("Group_Test_vs_Control", "TimePointd33.GroupTest")),
                                  alpha=0.05)
```

Number of genes with padj < 0.05 for Test v Control at day 11:

```{r}
sum(results.interaction.11$padj < 0.05, na.rm = TRUE)
```

Number of genes with padj < 0.05 for Test v Control at day 33:

```{r}
sum(results.interaction.33$padj < 0.05, na.rm = TRUE)
```

### Compare interaction model to the additive model with the LRT

Number of genes with padj < 0.05 for LRT of 
'~ TimePoint * Group' vs '~ TimePoint + Group'  

```{r, message=FALSE}
ddsObjC2 <- DESeq(ddsObj.interaction, test="LRT", reduced=additive.model)
resCSvCxS <- results(ddsObjC2)
sum(resCSvCxS$padj < 0.05, na.rm=TRUE)
```

Still not that many, but the results from the model for d11 and d33 differ 
by about 1700 genes.

## Differential gene expression analysis - Three groups (~SampleGroup)

From the clustering, PCA and correlation plots, the d11 and d33 controls do not
appear to be significantly different. This makes biological sense, they are just
mice that are a few week older.

How many genes have padj < 0.05 for d11 v d33 for Control:

```{r}
results.interaction.ctrl <- results(ddsObj.interaction, 
                                  name = "TimePoint_d33_vs_d11",
                                  alpha=0.05)
sum(results.interaction.ctrl$padj < 0.05, na.rm = TRUE)
```

What if we could merge the controls and treat them as a single group?

```{r DGE, message=FALSE}
samsht <- samplesheet %>% 
  mutate(SampleGroup = ifelse(Group=="Control",
                              "Control", 
                              str_c("Test.", as.character(TimePoint)))) %>% 
  mutate(SampleGroup = factor(SampleGroup))

grouped.model <- as.formula(~ SampleGroup)

ddsObj.raw <- DESeqDataSetFromTximport(txi = txi,
                                      colData = samsht,
                                      design = grouped.model)
keep <- rowSums(counts(ddsObj.raw)) > 5
ddsObj.filt <- ddsObj.raw[keep,]
ddsObj.grouped <- DESeq(ddsObj.filt)
results.grouped.11 <- results(ddsObj.grouped, 
                              name="SampleGroup_Test.d11_vs_Control", 
                              alpha=0.05)
results.grouped.33 <- results(ddsObj.grouped, 
                              name="SampleGroup_Test.d33_vs_Control", 
                              alpha=0.05)
```

Number of genes with padj < 0.05 for Test.d11 v Control:

```{r}
sum(results.grouped.11$padj < 0.05, na.rm = TRUE)
```

Number of genes with padj < 0.05 for Test.d33 v Control at day 33:

```{r}
sum(results.grouped.33$padj < 0.05, na.rm = TRUE)
```

# A plot of something

```{r}
gens <- c(rownames(results.interaction.11[which(results.interaction.11$padj<0.05),]),
          rownames(results.interaction.33[which(results.interaction.33$padj<0.05),])) %>% 
  unique()

dat <- counts(ddsObj.interaction, normalized=TRUE) %>% 
  as.data.frame() %>% 
  rownames_to_column("GeneID") %>% 
  filter(GeneID%in%gens) %>% 
  pivot_longer(names_to = "SampleName", values_to = "Exp", -GeneID) %>% 
  left_join(samplesheet) %>%
  group_by(Group, GeneID, TimePoint) %>% 
  summarize(MeanExp = mean(Exp)) %>%
  ungroup() %>% 
  mutate(MeanExp=log2(MeanExp))

dat2  <- dat %>% 
  pivot_wider(names_from = TimePoint, values_from = MeanExp) %>% 
  mutate(d33=d33-d11) %>% 
  mutate(d11=d11-d11) %>% 
  pivot_longer(names_to = "TimePoint", values_to = "MeanExp", starts_with("d"))


dat2 %>% 
  mutate(GeneGrp=str_c(GeneID, Group)) %>% 
  ggplot() +
    geom_point(aes(x=TimePoint, y=MeanExp, colour=Group)) +
    geom_line(aes(x=TimePoint, y=MeanExp, group=GeneGrp, colour=Group)) +
    facet_wrap(vars(Group))
```
